symforce/slam/imu_preintegration/manifold_symbolic.py

```diff
@@ -0,0 +1,346 @@
+]:
+    """
+    An internal helper function to update a set of preintegrated IMU measurements with a new pair of
+    gyroscope and accelerometer measurements. Returns the updated preintegrated measurements.
+
+    When integrating the first measurement, DR should be the identity, Dv, Dp, covariance, and the
+    derivatives w.r.t. to bias should all be 0.
+
+    Args:
+        DR (sf.Rot3): Preintegrated change in orientation
+        Dv (sf.V3): Preintegrated change in velocity
+        Dp (sf.V3): Preintegrated change in position
+        covariance (sf.M99): Covariance of [DR's tangent space, Dv, Dp]
+        DR_D_gyro_bias (sf.M33): Derivative of DR w.r.t. gyro_bias
+        Dv_D_gyro_bias (sf.M33): Derivative of Dv w.r.t. gyro_bias
+        Dv_D_accel_bias (sf.M33): Derivative of Dv w.r.t. accel_bias
+        Dp_D_gyro_bias (sf.M33): Derivative of Dp w.r.t. gyro_bias
+        Dp_D_accel_bias (sf.M33): Derivative of Dp w.r.t. accel_bias
+        gyro_bias (sf.V3): Initial guess for the gyroscope measurement bias
+        accel_bias (sf.V3): Initial guess for the accelerometer measurement bias
+        gyro_cov (sf.M33): The covariance of the gyroscope measurement
+        accel_cov (sf.M33): The covariance of the accelerometer measurement
+        gyro (sf.V3): The gyroscope measurement
+        accel (sf.V3): The accelerometer measurement
+        dt (T.Scalar): The time between the new measurements and the last
+        epsilon (T.Scalar): epsilon for numerical stability
+
+    Returns:
+        T.Tuple[sf.Rot3, sf.V3, sf.V3, sf.Matrix, sf.Matrix, sf.Matrix, sf.Matrix, sf.Matrix, sf.Matrix]:
+            new_DR,
+            new_Dv,
+            new_Dp,
+            new_covariance,
+            new_DR_D_gyro_bias,
+            new_Dv_D_gyro_bias,
+            new_Dv_D_accel_bias,
+            new_Dp_D_gyro_bias,
+            new_Dp_D_accel_bias
+    """
+    # Correct for IMU bias
+    corrected_accel = accel - accel_bias
+    corrected_gyro = gyro - gyro_bias
+
+    # Integrate the state
+    new_DR, new_Dv, new_Dp = integrate_state(
+        DR,
+        Dv,
+        Dp,
+        gyro=corrected_gyro,
+        accel=corrected_accel,
+        dt=dt,
+        epsilon=epsilon,
+    )
+
+    # NOTE(Brad): Both ways of calculating the derivatives of new_state should be the same.
+    if False:  # pylint: disable=using-constant-test
+        # Definitely correct, but not very amenable to CSE
+        def new_state_D(wrt_variables: T.List[T.Any]) -> sf.Matrix:
+            return sf.Matrix.block_matrix(
+                [
+                    tangent_jacobians(new_DR, wrt_variables),
+                    tangent_jacobians(new_Dv, wrt_variables),
+                    tangent_jacobians(new_Dp, wrt_variables),
+                ]
+            )
+
+        new_state_D_state = new_state_D([DR, Dv, Dp])
+        new_state_D_gyro_bias = new_state_D([gyro_bias])
+        new_state_D_accel_bias = new_state_D([accel_bias])
```

Comment:
This seems like a fantastic thing to compare against the handwritten version in a unit test :)

Reply:
Hmm. Yes. You're right. I'll come up with some way to make comparing them simple and clean.

Reply:
Note. I haven't actually gotten around to this. Not sure about it's priority on whether it should be punted indefinitely, reserved for a follow up PR, or added to this one (or the unit test PR that will be merged at the same time as this one).

Reply:
I guess I was hoping we'd do this in the unit test PR - it is a little unfortunate to just leave this here, pylint is right that constant tests are usually a bad sign.  If we're not going to add this test in this PR or the unit test PR we should probably at least make this an argument to this function that defaults to the handwritten version

Reply:
The only thing that had me hesitate was sorting out how to split out the two versions (auto-derivatives vs handwritten derivatives). The thing that made me hesitate about adding an argument to the function was that it would force us to use something like `functools.partial` to hide it from `Codegen`.

But, when I thought about it a bit more, I didn't see a better alternative. Anyway, since I made it possible to generate two separate versions, I went ahead and added the unit test to the unit test PR.